source/common/http/http1/codec_impl.cc

```diff
@@ -1133,20 +1135,22 @@ int ClientConnectionImpl::onHeadersComplete() {
       }
     }
 
-    if (parser_.status_code == 100) {
-      // http-parser treats 100 continue headers as their own complete response.
-      // Swallow the spurious onMessageComplete and continue processing.
-      ignore_message_complete_for_100_continue_ = true;
+    if (parser_.status_code == enumToInt(Http::Code::Continue)) {
       pending_response_.value().decoder_->decode100ContinueHeaders(std::move(headers));
-
-      // Reset to ensure no information from the continue headers is used for the response headers
-      // in case the callee does not move the headers out.
-      headers_or_trailers_.emplace<ResponseHeaderMapPtr>(nullptr);
     } else if (cannotHaveBody() && !handling_upgrade_) {
       deferred_end_stream_headers_ = true;
     } else {
       pending_response_.value().decoder_->decodeHeaders(std::move(headers), false);
     }
+
+    // http-parser treats 1xx headers as their own complete response. Swallow the spurious
+    // onMessageComplete and continue processing for purely informational headers.
+    if (CodeUtility::is1xx(parser_.status_code) &&
```

Comment:
Comment on why switching protocols is exempt (all data after headers is through-proxied as upgrade)